### ISSUE TITLE
Subsonic Provider Issues

### ISSUE THREAD

**Issue by rwlove** [music-assistant/hass-music-assistant#2192]
### What version of Music Assistant has the issue?

2.0.0b144

### What version of the Home Assistant Integration have you got installed?

N/A

### Have you tried everything in the Troubleshooting FAQ and reviewed the Open and Closed Issues and Discussions to resolve this yourself?

- [X] Yes

### The problem

There are a few observed errors when adding the Subsonic Provider.

1. Not all Artists/Albums/Tracks are imported.
* MA imports two less "album artists" than gonic lists, but many fewer than gonic lists as "artists".
* MA imported about 86% of the albums that gonic lists.
* MA imported about 0.6% of the tracks that gonic lists.

2. Some albums have missing album art (i.e. broken image), but more importantly the tripple-dot and the heart are greyed out, so they can't be interacted with or deleted.

3. When selecting one Album that was imported, MA could not list any tracks.

### How to reproduce

1. Completely clean install
2. Add Subsonic provider
3. Wait for synchronization to complete

### Music Providers

Subsonic

### Player Providers

None

### Full log output

2024-05-04 15:46:40.502 INFO (MainThread) [music_assistant] Starting Music Assistant Server (c3d4244d08ab4caa960bd02119589021) version 2.0.0b144 - HA add-on: False - Safe mode: False
2024-05-04 15:46:40.513 INFO (MainThread) [music_assistant.cache] Initializing cache controller...
2024-05-04 15:46:40.910 INFO (MainThread) [music_assistant.music] Using a sync interval of 180 minutes.
2024-05-04 15:46:41.304 INFO (MainThread) [music_assistant.streams] Detected ffmpeg version 6.0.1 with libsoxr support
2024-05-04 15:46:41.305 INFO (MainThread) [music_assistant.streams] Starting server on  0.0.0.0:8097 - base url: http://10.42.8.163:8097
2024-05-04 15:46:41.312 INFO (MainThread) [music_assistant.webserver] Starting server on  0.0.0.0:8095 - base url: http://10.42.8.163:8095
2024-05-04 15:46:41.507 INFO (MainThread) [music_assistant] Loaded metadata provider fanart.tv Metadata provider
2024-05-04 15:46:41.508 INFO (MainThread) [music_assistant] Loaded music provider Music Assistant
2024-05-04 15:46:41.508 INFO (MainThread) [music_assistant] Loaded metadata provider MusicBrainz Metadata provider
2024-05-04 15:46:41.508 INFO (MainThread) [music_assistant] Loaded metadata provider TheAudioDB Metadata provider
2024-05-04 15:46:41.686 INFO (MainThread) [music_assistant.music] Sync task for Music Assistant completed
2024-05-04 15:47:48.041 ERROR (MainThread) [music_assistant.webserver] Error handling message: config/providers/save: list index out of range
2024-05-04 15:47:59.723 INFO (MainThread) [music_assistant] Loaded music provider Open Subsonic Media Server Library
2024-05-04 15:47:59.880 WARNING (ThreadPoolExecutor-0_15) [py.warnings] /usr/local/lib/python3.12/site-packages/libopensonic/media/media_base.py:65: UserWarning: Artist object returned by server is missing required field 'sortName'
  warn(f"{self.get_class_name()} object returned by server is missing required field '{key}'")

2024-05-04 15:55:10.940 WARNING (ThreadPoolExecutor-0_5) [py.warnings] /usr/local/lib/python3.12/site-packages/libopensonic/media/media_base.py:65: UserWarning: Artist object returned by server is missing required field 'sortName'
  warn(f"{self.get_class_name()} object returned by server is missing required field '{key}'")

2024-05-04 15:55:37.940 WARNING (MainThread) [music_assistant.music] Sync task for Open Subsonic Media Server Library completed with errors
2024-05-04 15:58:39.032 WARNING (ThreadPoolExecutor-0_20) [py.warnings] /usr/local/lib/python3.12/site-packages/libopensonic/media/media_base.py:65: UserWarning: Artist object returned by server is missing required field 'sortName'
  warn(f"{self.get_class_name()} object returned by server is missing required field '{key}'")

2024-05-04 15:58:49.188 ERROR (MainThread) [music_assistant.webserver] Error handling message: music/albums/album_tracks: Field "disc_number" of type int in AlbumTrack has invalid value None
2024-05-04 15:59:38.972 ERROR (MainThread) [music_assistant.webserver] Error handling message: music/albums/album_tracks: Field "disc_number" of type int in AlbumTrack has invalid value None

### Additional information

I am on Discord as 'Bubba' and I'm happy to chat and help test/debug.

My container output shows a few lines of `al-2628: No such file or directory` that I don't see in the logfile.

### What version of Home Assistant Core are your running

N/A

### What type of installation are you running?

Home Assistant Container

### On what type of hardware are you running?

Alternative

**Comment by OzGav** [comment]
@khers

**Comment by khers** [comment]
I saw this behavior in my HA install and have been trying to investigate. I cannot reproduce in my dev env, there the scan completes successfully and when I update to beta 147 it seems to be fixed for me there. Please try the new beta and let me know if it is still broken.

**Comment by khers** [comment]
It looks like this took 2 hours to sync. Do you see incorrect results still? Once my sync finally finished the stats are what I expect (MA and Gonic disagree about the number of albums and artists because my metadata is not the cleanest.

**Comment by khers** [comment]
For me, the ghost entries are usually a misspelling of a band name or artist name (or album). I intend to poke around in my test instance db and see if I can find a pattern to what is causing these entries and recommend metadata fixes for it.

IMO this issue is caused by MA and Gonic disagreeing about how metadata works and the relationships between entities, and the provider code does not include any massaging of these relationships. I don't think fixing it in provider code is the correct choice because this is an Open Subsonic provider and not a Gonic provider.

**Comment by OzGav** [comment]
@rwlove As there has been no followup to the above we will close this soon.

**Comment by khers** [comment]
Sorry, I didn't mean to sound like this is a "Won't Fix" more that it is going to require some more investigation to get to the bottom of what is happening here. I hope to spend some time looking through the MA database and compare to what is in my Gonic collection.

**Comment by khers** [comment]
I don't have any insight into the cause of the missing tracks or albums but I am fairly certain that the extra artists are a disagreement between the artist as constructed by Gonic when it scans the filesystem and the entries in the Artist field of the ID3 tags. AFAIK Gonic does not build new artist entries for what it finds in the ID3 tags (I'd have to ask the  maintainer to verify) but when we parse each track in MA we make sure to create an Artist entry for each value found in the ID3 tags and for each item returned by the track enumeration. I have been putting off the clean up of my metadata but I think I am at the point where it is the only way forward for this investigation. I will try and get the motivation up for that project in the next week or so.

**Comment by khers** [comment]
So I finally got around to cleaning up my meta data with beets and the difference from gonic to MA is larger?

`type                   gonic        MA
artists                1361         1443
album artists    654            N/A
albums              1206         1148
tracks                 12254      11538`

I am at a loss for how to explain the discrepancy between albums and tracks, these should be 1-1. All I can think of for the artists is that gonic counts album artists as unique names in that field of the metadata for any track, but I suspect there are duplicates between artists and album artists. I wonder if the union between artists and album artists in gonic would match what MA sees.

**Comment by khers** [comment]
I am asking the gonic maintainer some questions about what might be going on here: https://github.com/sentriz/gonic/issues/534

**Comment by khers** [comment]
@rwlove Do you have examples of "ghost" entries. I think the next step here is to look deeper into what each service believe they have and see where the discrepancies are.

**Comment by khers** [comment]
I think what I need to do is dump the tracks from my Gonic install and then from the MA database and remove the intersection to see what tracks MA excluded. This should work for aartists and albums as well. I will let you know what I find when I have the motivation to do that.

**Comment by OzGav** [comment]
@khers any progress on this one?

**Comment by khers** [comment]
Given that this isn't actually harming anyones ability to use the provider (and $DAYJOB is going nuts) I won't be able to spend a ton of time investigating this. If the project prefers issues to be closed, I am fine to close this one and make a comment in the code about this issue for future investigation.

**Comment by OzGav** [comment]
This can stay open. Is there a known issue I should add to the docs?

**Comment by khers** [comment]
We should probably mention that having a mismatch between the counts of artists, albums, and tracks is a possibility and link back to this issue. I'd love it is someone showed up with a specific example of a missing or ghost track.

**Comment by rfporter** [comment]
Has anyone else noticed that compilation tracks sometimes get the wrong artist assigned to them. The album artist is "Various Artists" then each track has its own artist. The track artists are correct in Navidrome and Symfonium, but in MA some of them have been set to James Horner (in my setup at least).

I think that this happens if the track artist does not match any existing Album Artist in the library.

**Comment by OzGav** [comment]
Sounds like this https://github.com/music-assistant/hass-music-assistant/issues/2965

**Comment by rfporter** [comment]
> Sounds like this #2965

Looks like it. However its not only for Multi-artist tracks. It seems to be for any track by an artist that does not have an album in the library.

**Comment by khers** [comment]
This is due to the way the Navidrome devs represent these artists. They do not have an entry in the artists table so MA sees them for the first time during track scanning. I thought I had gotten the track artists working with the last hack I put in place to give these artists an ID that we could use internally but it's possible that that hack is broken in this case. As I stated in the other issue, I do not run a Navidrome server so testing against one is a challenge for me. I run Gonic so that gets well tested.

As with the other bug there are two possible solutions:
1. Someone here who has a Navidrome server they can test against puts together changes to the subsonic provider that I can verify doesn't break other implementations.
2. Someone convinces the Navidrome devs that their implementation of multi-artist albums and artists without an album credit needs to change

The first is a far easier path unless you want to become a Navidrome dev.

**Comment by khers** [comment]
As I said in #2965 I have a PR that might address the multi-artists album issue: https://github.com/music-assistant/server/pull/1758 once that is merged (there are ruff issues I need to address) can I get a Navidrome user to try it and report back?

**Comment by khers** [comment]
:( That's the problem with trying to dash off a fix without a proper test, thanks for the catch (and for testing). I have updated the PR (and cut a new libopensonic to fix).

**Comment by OzGav** [comment]
@rwlove We will close this soon assuming fixed

**Comment by rwlove** [comment]
@OzGav I'll close it. Thanks for the help.
